[PGNCCL] Add FP8 support

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152706

NCCL added support for `Float8e4m3` and `Float8e5m2` in 2.24.

NVIDIA GPUs does not seem to support the following "no negative zero" versions: `Float8_e4m3fnuz` and `Float8_e5m2fnuz`, see https://onnx.ai/onnx/technical/float8.html. So we continue to error out for these two upon a reduction op.

Test plan:
- test_allreduce_float8
- test_reduce_scatter_float8

Resolves #148344

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 committed:
[PGNCCL] Add FP8 support

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4110,7 +4129,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::allreduce_sparse(
   TORCH_CHECK(
-      !isFloat8Type(tensor.scalar_type()),
+      !isUnsupportedFloat8(tensor.scalar_type()),
       "Float8 dtypes are not currenlty supported for NCCL reductions");
```
should we changes these comments to indicate which types are supported?

applies below as well

kwen2501 replied:
Good point.

fduwjj replied:
Also is isFloat8Type still being used? If not, maybe you want to consider to remove it as well?

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4110,7 +4129,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::allreduce_sparse(
   TORCH_CHECK(
-      !isFloat8Type(tensor.scalar_type()),
+      !isUnsupportedFloat8(tensor.scalar_type()),
       "Float8 dtypes are not currenlty supported for NCCL reductions");
```
nit: spelling "currently"

applies below as well

kwen2501 replied:
Yeah, there is a separate fix, ignoring here to avoid a potential conflict.

d4l3k reviewed (approved):
Can we add some unit tests for these types as well? Do we have fp8 GPUs available in CI?

kwen2501 commented:
> Can we add some unit tests for these types as well? Do we have fp8 GPUs available in CI?

The requirement for FP8 is SM >= 90 (Hopper).
Distributed tests are using T4.

eqy commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4230,7 +4249,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::allreduce(
   }
   TORCH_CHECK(
-      !isFloat8Type(tensor.scalar_type()),
+      !isUnsupportedFloat8(tensor.scalar_type()),
```
Not from this PR but maybe fix the obvious typos "currenlty" -> "currently" haha

fduwjj reviewed (approved):
Opps forgot to send out stamp

kwen2501 committed:
Update on "[PGNCCL] Add FP8 support"


NCCL added support for `Float8e4m3` and `Float8e5m2` in 2.24.

NVIDIA GPUs does not seem to support the following "no negative zero" versions: `Float8_e4m3fnuz` and `Float8_e5m2fnuz`, see https://onnx.ai/onnx/technical/float8.html. So we continue to error out for these two upon a reduction op.

Resolves #148344

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Add FP8 support"


NCCL added support for `Float8e4m3` and `Float8e5m2` in 2.24.

NVIDIA GPUs does not seem to support the following "no negative zero" versions: `Float8_e4m3fnuz` and `Float8_e5m2fnuz`, see https://onnx.ai/onnx/technical/float8.html. So we continue to error out for these two upon a reduction op.

Resolves #148344

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Add FP8 support"


NCCL added support for `Float8e4m3` and `Float8e5m2` in 2.24.

NVIDIA GPUs does not seem to support the following "no negative zero" versions: `Float8_e4m3fnuz` and `Float8_e5m2fnuz`, see https://onnx.ai/onnx/technical/float8.html. So we continue to error out for these two upon a reduction op.

Resolves #148344

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Add FP8 support"


NCCL added support for `Float8e4m3` and `Float8e5m2` in 2.24.

NVIDIA GPUs does not seem to support the following "no negative zero" versions: `Float8_e4m3fnuz` and `Float8_e5m2fnuz`, see https://onnx.ai/onnx/technical/float8.html. So we continue to error out for these two upon a reduction op.

Test plan:
- test_allreduce_float8
- test_reduce_scatter_float8

Resolves #148344

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

Skylion007 commented:
@pytorchbot merge
